### PR TITLE
Document the lifespan event handler parameter

### PR DESCRIPTION
**Pull request by emilmelnikov**
https://github.com/encode/starlette/pull/799 added support for the `lifespan` parameter, but it was not reflected in the documentation.

### CODE REVIEW

File: docs/events.md [hunk 1]
```diff
@@ -37,6 +37,25 @@ registered startup handlers have completed.
 The shutdown handlers will run once all connections have been closed, and
 any in-process background tasks have completed.
 
+A single lifespan asyncronous generator handler can be used instead of
```
Comment (graingert) [inline docs/events.md]:
```suggestion
A single lifespan asynccontextmanager handler can be used instead of
```

**Review by graingert — CHANGES_REQUESTED** [review]
lifespan now takes an async context manager factory, eg an async generator function decorated with `@contextlib(2).asynccontextmanager`

File: docs/events.md [hunk 2]
```diff
@@ -37,6 +37,25 @@ registered startup handlers have completed.
...
+separate startup and shutdown handlers:
+
+```python
+from starlette.applications import Starlette
+
+
+async def lifespan(app):
```
Comment (graingert) [inline docs/events.md:50]:
```python
@contextlib.asynccontextmanager
async def lifespan(app):
    # acquire async resources
    async with anyio.create_task_group() as app.tg:
        try:
            yield
        finally:
            with anyio.CancelScope(shield=True):
                # release async resources
```

File: docs/events.md [hunk 3]
```diff
@@ -37,6 +37,33 @@ registered startup handlers have completed.
...
+
+@contextlib.asynccontextmanager
+async def lifespan(app):
+    # acquire async resources
+    async with anyio.create_task_group() as app.tg:
+        try:
+            yield
+        finally:
+            with anyio.CancelScope(shield=True):
+                # release async resources
```
Comment (graingert) [inline docs/events.md]: sorry I lead you wrong here - the TaskGroup context manager will capture any CancelledError for you

```suggestion
    async with anyio.create_task_group() as app.tg:
        yield
    # release async resources
```

Comment (graingert) [reply docs/events.md]: I do want to somehow document that `yield` can now throw CancelledError

Comment (graingert) [reply docs/events.md]: sometimes you want to acquire/release resources inside the TaskGroup and sometimes you want to acquire/release them outside the TaskGroup and there's no specific answer as they both have different behaviours with their own pros/cons

image

Comment (uSpike) [reply docs/events.md]: I'm not sure we should have the main documentation for lifespan be setting `app.tg`.  That could be a separate topic, but I think the initial introduction to "make a lifespan" example should be as simple as possible.

```python
@contextlib.asynccontextmanager
async def lifespan(app):
    async with some_async_resource():
        yield
```

Comment (graingert) [reply docs/events.md]: that's fair

**Comment by graingert** [comment]
Blocks https://github.com/encode/starlette/pull/1233

File: docs/events.md (hunk 3 above)
Comment (emilmelnikov) [reply docs/events.md]: I like the simple lifespan example; we can also add a link to [anyio documentation](https://anyio.readthedocs.io/) on task groups and cancellation.

**Review by graingert — APPROVED** [review]

**Comment by graingert** [comment]
Thanks! Apologies for dragging this one off topic in my original review

**Pull request merged**